runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/status/BeamWorkerStatusGrpcService.java

```diff
@@ -0,0 +1,207 @@
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (CompletableFuture<WorkerStatusClient> clientFuture : connectedClient.values()) {
+      if (clientFuture.isDone()) {
+        clientFuture.get().close();
+      }
+    }
+    connectedClient.clear();
+  }
+
+  @Override
+  public StreamObserver<WorkerStatusResponse> workerStatus(
+      StreamObserver<WorkerStatusRequest> requestObserver) {
+    String workerId = headerAccessor.getSdkWorkerId();
+    LOG.info("Beam Fn Status client connected with id {}", workerId);
+
+    WorkerStatusClient fnApiStatusClient =
+        WorkerStatusClient.forRequestObserver(workerId, requestObserver);
+    if (connectedClient.containsKey(workerId) && connectedClient.get(workerId).isDone()) {
+      LOG.info(
+          "SDK Worker {} was connected to status server previously, disconnecting the old client",
+          workerId);
+      try {
+        WorkerStatusClient oldClient = connectedClient.get(workerId).get();
+        oldClient.close();
+      } catch (IOException | InterruptedException | ExecutionException e) {
+        LOG.warn("Error closing worker status client", e);
+      }
+    }
+    connectedClient
+        .computeIfAbsent(workerId, k -> new CompletableFuture<>())
+        .complete(fnApiStatusClient);
+    return fnApiStatusClient.getResponseObserver();
+  }
+
+  /**
+   * Get the latest SDK worker status from the client's corresponding SDK Harness.
+   *
+   * @param workerId worker id of the SDK harness.
+   * @return {@link CompletableFuture} of Status info from SDK harness if successful, otherwise
+   *     returns an error message.
+   */
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public CompletableFuture<String> getWorkerStatus(String workerId) {
+    try {
+      CompletableFuture<WorkerStatusResponse> workerStatus =
+          getStatusClient(workerId).getWorkerStatus();
+      CompletableFuture<String> result = new CompletableFuture<>();
+      workerStatus.thenAcceptAsync((r) -> result.complete(getStatusString(r)));
+      return result;
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      LOG.debug(String.format("Error getting worker status from %s", workerId), e);
+      return CompletableFuture.completedFuture(DEFAULT_ERROR_RESPONSE);
+    }
+  }
+
+  /**
+   * Get all the statuses from all connected SDK harnesses within specified timeout.
+   *
+   * @param timeout max time waiting for the response from all SDK harness.
+   * @param timeUnit timeout time unit.
+   * @return All the statuses in a Map which key is the SDK harness id and value is the status
+   *     content.
+   */
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public Map<String, String> getAllWorkerStatuses(long timeout, TimeUnit timeUnit)
+      throws InterruptedException {
+    // return result in worker id sorted map.
+    Map<String, String> allStatuses =
+        new ConcurrentSkipListMap<>(
+            (first, second) -> {
```

Comment:
Shall we simply rely on natural ordering of string here?

Reply:
I think it might not be as user-friendly when using natural ordering, eg. it is a little bit weird when rendering page with the order 'sdk1 sdk10 sdk11 ... sdk2 sdk20 ...' but I don't have strong opinion on this.

Reply:
We don't follow sdk naming structure so adding compactor to utilize this assumption might not be very useful.

Reply:
I see.